core/network/impl/synchronizer_impl.cpp

```diff
@@ -1406,4 +1408,81 @@ namespace kagome::network {
     ancestry_.clear();
     recent_requests_.clear();
   }
+
+  bool SynchronizerImpl::fetchHeadersBack(primitives::BlockNumber max, primitives::BlockNumber min,
+    CbResultVoid cb) {
+    if (max < min) {
+      SL_ERROR(log_, "max {} < min {}, will not fetch headers", max, min);
+      return false;
+    }
+
+    BlocksRequest request{
+        .fields = BlockAttribute::HEADER,
+        .from = max,
+        .direction = Direction::DESCENDING,
+        .max = max - min + 1,
+        .multiple_justifications = false,
+    };
+    auto chosen = chooseJustificationPeer(max, request.fingerprint());
+    if (not chosen) {
+      return false;
+    }
+
+    auto cb2 = [weak{weak_from_this()},
+                cb{std::move(cb)},
+                peer{*chosen}](outcome::result<BlocksResponse> r) mutable {
+      auto self = weak.lock();
+      if (not self) {
+        return;
+      }
+
+      self->busy_peers_.erase(peer);
+      if (not r) {
+        return cb(r.error());
+      }
+
+      auto &blocks = r.value().blocks;
+      if (blocks.size() == 0) {
+        return cb(Error::EMPTY_RESPONSE);
+      }
+      std::optional<primitives::BlockHash> parentHash;
+      for (auto& b : blocks) {
+        auto &header = b.header;
+
+        if (not header) {
+          return cb(Error::EMPTY_RESPONSE);
+        }
+
+        auto& headerValue = header.value();
+        primitives::calculateBlockHash(headerValue, *self->hasher_);
+        const auto& headerInfo = headerValue.blockInfo();
+
+        if (self->block_tree_->isFinalized(headerInfo)) {
+          parentHash = headerValue.parent_hash;
+        } else {
+          if (not parentHash.has_value()) {
+            SL_WARN(self->log_, "parentHash is not set yet and block #{} is not finalized, will not be stored", headerInfo.number);
+            continue;
+          }
+          if (*parentHash != headerInfo.hash) {
+            SL_ERROR(self->log_, "Parent hash mismatch in block #{}", headerInfo.number);
+            return cb(Error::INVALID_HASH);
+          }
+          parentHash = headerValue.parent_hash;
+        }
+
+        if (auto er = self->block_storage_->putBlockHeader(headerValue); er.has_error()) {
+          SL_ERROR(self->log_, "Failed to put block header: {}", er.error());
+          return cb(er.error());
+        }
+        if (auto er = self->block_storage_->assignNumberToHash(headerInfo); er.has_error()) {
+          SL_ERROR(self->log_, "Failed to assign number to hash: {}", er.error());
+          return cb(er.error());
+        }
```

Comment:
```c++
        if (header->blockInfo() != expected) {
          return cb(Error::INVALID_HASH);
        }
        putBlockHeader(...)
        if (finalized) {
          assignNumberToHash(...)
        }
        auto parent = header->parentInfo();
        if (not parent) {
          break;
        }
        expected = *parent;
```

Reply:
Why should we hold the whole info, not just hash?

Reply:
Ok, let's check just hash